aws-sdk integration

ericmustin opened the pull request:
### What does this PR do?

_Note: I wanted to get this in early though it's not quite there yet. tl;dr this is a lot of what I think this integration needs to look like but was hoping for some feedback on instrumentation approach/robustness and some broader tagging/naming conventions._

This PR adds automatic instrumention for [aws-sdk-js](https://github.com/aws/aws-sdk-js/), to improve granularity around outgoing http requests to different AWS Services. These are collected under a span with operation name `aws.http` and include instrumentation for `AWS.Service` request either: 
  - with callbacks
  - with `.send()`
  - with `.promise()`

- There is tagging currently on all spans for:
  - `aws.agent`
  - `aws.operation`
  - `aws.request_id`
  - `aws.region`
  - `aws.service`
  - `aws.url`
  - `component`
  - `http.status_code`
  - `http.content_length`


- For the following services there are extra tags:
  - DyanamoDB: `aws.dynamo_db.table_name`
  - Kinesis: `aws.kinesis.stream_name`
  - S3: `aws.s3.bucket_name`
  - SQS: `aws.sqs.queue_name`
  - SNS: `aws.sns.topic_arn`

These tags follow the conventions established in serverless team instrumentation / conversion of x-ray trace details, based on [snapshots seen here](https://github.com/DataDog/dd-go/tree/prod/cloud-integrations/internal/aws/xray-converter/request/testdata )

The resource naming convention also follows the pattern established by the serverless team of [`<operation_name> <table/bucket/queue/stream/topic name>` ](https://github.com/DataDog/dd-go/blob/a22c369566cc4274c49a52fe318ad8f38a8ddfee/cloud-integrations/internal/aws/xray-converter/request/testdata/dynamodb_segments.json~snapshot#L95) or just `<operation_name>` in the case that the user is interacting with a service which does not have extra granularity.

Additionally this PR includes a configuration option `config.hooks.addCustomTags` which is a function with arguments of the curent span and the request parameters `(span, params) => {}`, so that user's can add tags for a parameter in the request body that is not picked up automatically.

Notes / Questions / Todos:
  - Todos:
    - The request_id appears to be inconsistently returned. All it is is the `x-amz-request-id` response header, so it seems possible to me that the localstack test container is malforming or failing to set this in certain cases. I'm not sure if it just has some actual inconsistency or if it's specific to the test env. I plan on testing this against some production aws endpoints to confirm, and if that is the case I can look into forcing an `x-amz-request-id` response header on all responses in the test. In the meantime I've commented out this assertion in tests where it was flaking.
    - The same goes for `http.status_code` and `http.content_length` on error responses from localstack, the `response.httpResponse` key is not generated correctly. I need to confirm whether this is specific to the mocks returned or if this is an actual aws behavior. again in the meantime i've commented out these assertions.
    - The tests in general could be a little cleaner on some of the services, didn't feel blocking from getting a pr on to get review process of main instrumentation started, but I plan to clean this up a bit more.

    - If there are other main aws services or metadata(is there any sql querys to capture?) that are worth adding in the initial bit here, and not something added later if a user has interest in it, let me know and i can add.

  - Questions / Notes:
    - currently the service name is one specific `app-aws` or `<service-name>-aws`. The resource naming convention is only "operation" + "table/bucket/queue/etc name" in situations where it's available. However, the underlying service (dynamodb, sqs, chime, etc etc) is only specified by the `aws.service` and `component` tags, which is how this is specified in the RFC and the current format of x-ray converted traces. So in order to see latency of all "dynamoDB" spans it would require app analytics. However, I think it's a bit confusing, arguably users would want to see a service page and related metrics for "dynamodb" or "s3" and not have to rely on app analytics for this, while 1 general aws service would, imo, lose a lot of value since it combines some very low latency endpoints with potentially high latency ones. While you could still break it out by resource, groups of resources specific to s3, sqs, etc would not have their own service page and associated metrics. On the other hand, there are, at quick glance, dozens and dozens of potential AWS Services, and so it might explode the services list in an unintended way for some users.
    
    - hypothetically, could the `addCustomTags` hook be more useful here, and if the full AWS.Response object context is passed in(which is available in the [complete event or the callback](https://github.com/awsdocs/aws-javascript-developer-guide-v2/blob/master/doc_source/using-a-callback-function.md#accessing-the-request-and-response-objects), a user could hypothetically use this hook to update the service name so that they had a seperate "s3-app-aws" service, or something along those lines.
  
    - The AWS.Request object has hooks that allow arbitrary number of functions to be called on "send" that emit a [`complete`](https://github.com/awsdocs/aws-javascript-developer-guide-v2/blob/master/doc_source/using-a-response-event-handler.md#the-complete-event) event, which is how I'm currently instrumenting requests that do not have a [`callback`](https://github.com/awsdocs/aws-javascript-developer-guide-v2/blob/master/doc_source/using-a-callback-function.md#using-an-anonymous-callback-function) passed in explicitly. I think this approach is robust enough so that it won't lead to any strange async behavior from promises, but let me know if i've missed something.

### Plugin Checklist
<!-- Fill this section if adding or updating a plugin. Remove otherwise. -->

- [x] Unit tests.
- [x] TypeScript [definitions][1].
- [x] TypeScript [tests][2].
- [x] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [x] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/dd-trace-js/blob/master/docs/API.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Additional Notes
<!-- Anything else we should know when reviewing? -->

ericmustin commented on packages/datadog-plugin-aws-sdk/src/aws_helpers.js:
```diff
@@ -0,0 +1,159 @@
+      awsHelpers.finish(span, err, config)
+
+      if (typeof done === 'function') {
+        tracer.scope().activate(parent, () => {
```
should i be using tracer.scope.bind here? or more generally?

ericmustin replied:
i went ahead and updated both this section, and other relevant areas in `base.js`, to try to use `scope().bind()` instead of `scope().activate`. I _think_ that's the proper way to do instrumentation in this case? lmk what you think.

rochdev replied:
In general, `bind` would be used when you return the function without calling it, and `activate` is used when you call the function immediately. `bind` does work in both cases though, but it creates an additional function in the process compared to `activate`.

rochdev commented on packages/dd-trace/test/setup/services/aws_sdk.js:
```diff
@@ -0,0 +1,42 @@
+'use strict'
```
File is named `aws_sdk` but service is named `localstack`. They should match to make it easier to understand.

ericmustin replied:
makes sense, updated. frankly i'm not entirely clear where this file is getting run? It's used to validate that the container for these tests is listening on the right host/port? i used the gcloud-pub-sub instrumentation skeleton as a guide here but not sure i fully grok the use case

rochdev commented on packages/datadog-plugin-aws-sdk/src/base.js:
```diff
@@ -0,0 +1,91 @@
+
+      analyticsSampler.sample(span, config.analytics)
+
+      // https://github.com/awsdocs/aws-javascript-developer-guide-v2/blob/
```
Why this comment?

rochdev replied:
Same question for other similar comments.

ericmustin replied:
most of this was just to help with code review so everyone is familiar with what the args are , i can remove these "pointing to aws doc reference" type comments if its just messy and not helpful. I did add some code comments related to choices on _why_ the instrumentation is handling `with callback` and `without callback` differently, that i think are useful, but again if the conventions here is to keep the comments to a minimum , happy to remove.

rochdev replied:
The idea is mostly to not pollute the code with unnecessary comments, otherwise you unconsciously start to ignore them, similar to apps with a lot of warnings. When there is a comment, it should be required to understand why the code works a specific way, but in this case I felt like we're just describing AWS general best practices which is not that helpful. But I also wanted to make sure I didn't miss something important that would make this comment really needed.

rochdev commented on packages/datadog-plugin-aws-sdk/src/base.js:
```diff
@@ -0,0 +1,91 @@
+        const boundAwsReq = tracer.scope().bind(awsReq, span)
+
+        boundAwsReq.on('send', response => {
+          tracer.scope().activate(span)
```
This is not doing anything since a callback is not passed.

ericmustin replied:
good catch, removed unneeded argument for this listener. I also moved this listener to an earlier event `validate`, adding a comment in a moment discussing this

rochdev commented on packages/datadog-plugin-aws-sdk/src/base.js:
```diff
@@ -0,0 +1,91 @@
+'use strict'
```
What is the reasoning behind the current file structure? I'm not sure it helps readability unless it would be split more specifically.

ericmustin replied:
updated per https://github.com/DataDog/dd-trace-js/pull/856#discussion_r381882631

rochdev commented:
> On the other hand, there are, at quick glance, dozens and dozens of potential AWS Services, and so it might explode the services list in an unintended way for some users.

The way we currently do services this is how we should do it. If the app uses 20 different AWS services, then we should see them as 20 different services in the UI. We can revisit when this has been improved.

> hypothetically, could the addCustomTags hook be more useful here, and if the full AWS.Response object context is passed in

In general hooks should be passed the relevant information that was available at the time it was called. If an AWS request always returns a `AWS.Response` object, I would say this is probably the safe one to use.

> The AWS.Request object has hooks that allow arbitrary number of functions to be called on "send" that emit a complete event, which is how I'm currently instrumenting requests that do not have a callback passed in explicitly

Is the event always sent? If so, why not simply always use the event? Also, is it the same event for errors?

ericmustin commented on docker-compose.yml:
```diff
@@ -73,3 +73,14 @@ services:
+        - "127.0.0.1:4575:4575"
+      environment:
+        - LOCALSTACK_SERVICES=dynamodb:4569,kinesis:4568,s3:4572,sqs:4576,sns:4575
+        - EXTRA_CORS_ALLOWED_HEADERS=x-amz-request-id,x-amz-id-2
```
updated

ericmustin commented on docs/test.ts:
```diff
@@ -124,8 +124,16 @@ const elasticsearchOptions = {
+const awsSdkOptions = {
+  service: 'test',
+  hooks: {
+    addCustomTags: (span, params) => {},
```
Ok understood, that makes more sense to me now. I've updated this and also (regarding later comment) am just passing in the full AWS.Response Object. This will give users access to the response, response.request and response.request.params so they should be able to add any relevant metadata tags via this hook.

ericmustin commented on packages/datadog-plugin-aws-sdk/src/aws_helpers.js:
```diff
@@ -0,0 +1,191 @@
+'use strict'
```
gotcha, yea a little overkill on my part, tried to clean everything up and simplify/improve the naming conventions where i could, lmk if it looks alright

ericmustin commented on packages/datadog-plugin-aws-sdk/src/aws_helpers.js:
```diff
@@ -0,0 +1,191 @@
+
+  finish (span, err) {
+    if (err) {
+      span.addTags({
```
updated

ericmustin commented on packages/datadog-plugin-aws-sdk/src/aws_helpers.js:
```diff
@@ -0,0 +1,191 @@
+  },
+
+  addAwsResponseTags (span, response) {
+    if (span) {
```
agreed, just a force of habit for me. I've made a pass through `helpers` and tried to clean up the nesting where possible.

ericmustin commented on packages/datadog-plugin-aws-sdk/src/aws_helpers.js:
```diff
@@ -0,0 +1,191 @@
+    }
+  },
+
+  addAwsResponseTags (span, response) {
```
i made a first pass at re-org'ing all this tagging stuff, basically here's how it's now organized

1. Base Tags: mostly hardcoded values or things known before request time (passed to span constructor)
	- span type, span kind, service name, aws.region/operation/agent/url, etc

2. response tags: tags using the aws.response object (called just before a span finishes)
  	- This has seperate functions for
		- ServiceTags (resource tags + all the tags for specific aws services)
			- seperate functions for each aws service
		- RequestId Tag (request id is a weird one, with some breaking changes between versions, so i split it out into its own function)
		- Http Response Tags (details about http response like status code, content length)
		- custom Tags `config.hooks.http` , optional user config to collect arbitrary info

Technically there's some tags added in the "response" section that we would know at or before request time, but i wanted to keep all the service related tagging grouped together , and some of that does (or may likely in the future) depend on specific api response info. Also the AWS.Response object contains a `request` key that references the request so it's not much lift to access the request info from the response.

Lmk what you think

ericmustin commented on packages/datadog-plugin-aws-sdk/src/aws_helpers.js:
```diff
@@ -0,0 +1,191 @@
+    }
+  },
+
+  addAwsServiceTags (span, operation, params, serviceName) {
```
yup, agreed, updated, lmk what u think

ericmustin commented on packages/datadog-plugin-aws-sdk/src/aws_helpers.js:
```diff
@@ -0,0 +1,191 @@
+    // TODO: move to case statement
+    if (operation && params) {
+      switch (serviceName) {
+        case 'Amazon.DynamoDB':
```
i think we can probably centralize discussion here, https://github.com/DataDog/dd-trace-js/pull/856#discussion_r381875624

 but yes agreed

ericmustin commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,5 @@
+'use strict'
+
+module.exports = [].concat(
```
removed, originally i had thought i might need multiple files for patching different services, but that has turned out not to be the case.

ericmustin commented:
@rochdev still answering one or two feedback points but re block comment responses:

--------------------
> The way we currently do services this is how we should do it. If the app uses 20 different AWS services, then we should see them as 20 different services in the UI. We can revisit when this has been improved.

Ok, I've updated this to append the aws service to the service name, so

`myApp-Amazon.S3` or `myApp-Amazon.DynamoDB` etc

--------------------
> In general hooks should be passed the relevant information that was available at the time it was called. If an AWS request always returns a AWS.Response object, I would say this is probably the safe one to use.

Makes sense, I've updated this so the args are `(span, <AWS.Response Object>) => {}`

--------------------
> Is the event always sent? If so, why not simply always use the event? Also, is it the same event for errors?

I spent some time digging into this because I was expecting this to _just work_ if i just always used the event and it didn't.

So, theres basically two paths here: `makeRequest` invoked with a callback and without a callback. 

With a callback, the constraint here is that `makeRequest` will just automatically [invoke `.send` if it detects a callback](https://github.com/aws/aws-sdk-js/blob/9c191bbdbf32a8a3fa31219e369006f852318a1f/lib/service.js#L202-L206) and then [.send will kick off the state machine (_yup, there's a state machine, more on this in a sec_)](https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L359-L370) before we have a chance to attach our listeners. So, we need to activate the span ourselves and then pass in a shimm'd callback to determine when the request has completed.

On the other hand, without a callback, the request object is returned without invoking `.send()` to kick off the state machine. So we can bind the AWS.Request to the scope and then start a span on the appropriate events on the state machine. 

The state machine itself is a bit complex, [seen here](https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L44-L57), but it turns out that `.on('send')` will state will _sometimes_ get skipped if there are earlier state machine errors(like on validation). This is reasonable imo, it avoids sending requests on _definitely malformed_ parameters. Moreover, it seems to me like we ought to be instrumenting as much of the state machine as possible since the work done here is both non-trivial timing/performance wise , and related to the aws service (and would provide deeper granularity into timing than what we'd get with just the http instrumentation). As best I can tell there are 2 events that are _always_ going to be called in the request cycle, the first ([`validate`](https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L142)) and last ([`complete`](https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L245)), so those are the ones where we `startSpan` and `finish` span, respectively. 

That was a lot, but tl;dr i think we need to two different instrumentation paths here as it's the most robust way i can think of to ensure we're always properly timing the spans

ericmustin commented on packages/datadog-plugin-aws-sdk/src/aws_helpers.js:
```diff
@@ -0,0 +1,191 @@
+    }
+  },
+
+  // names are inconsistently prefixed with AWS or Amazon
```
Fwiw this was just something the serverless folks mentioned in RFC comments, i haven't really vetted this fully. Seems like a very aws thing to do though. 

Assuming the inconsistency is frequent...In my mind, Given that this instrumentation is going to split out services by aws service, it would seem like we want to offer a way for users to also easily _group by_ related aws services. Searching in `traces/App analytics` allows things like [wildcarding](https://docs.datadoghq.com/logs/explorer/search/#wildcards) so adding some consistency to the prefix would make this much easier.

I also realize, at a high level, that tracing instrumentation shouldn't necessarily be written to fit an arbitrary UI convention, and there's also  probably _other_ indirect ways of grouping by aws service in the UI (say, searching for only traces with an `aws.agent` tag), but this just seemed like a low hanging fruit way to make things a little nicer. Happy to axe if it's more hassle than it's worth.

rochdev replied:
As discussed yesterday, there is no prefix anymore since we use a different field.

ericmustin commented on packages/datadog-plugin-aws-sdk/src/base.js:
```diff
@@ -0,0 +1,91 @@
+module.exports = [
+  {
+    name: 'aws-sdk',
+    versions: ['>=2.3.0'],
```
So 2.3.0 introduced the .promise() function which is the suggested way to use the package now, this was in `03/2016`.

  - https://aws.amazon.com/releasenotes/release-aws-sdk-for-javascript-v2-3-0/?tag=releasenotes%23keywords%23javascript
  - https://github.com/aws/aws-sdk-js/releases/tag/v2.3.0

And additionally, I think the main usage of this would be in Lambda's like here:

  - https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html

Which use `2.585.0` for both Node.js 10 and 12

However, for `>=2.0` I think mainly I would just need to do use `util.promise` instead of `promise()` for scaffolding in the before/after Each hook of the tests and then only test the `.promise()` with specific versions. Not a huge lift but just will need to implement, i don't think anything changes regarding instrumentation. I can get started on that.

For 1.x support, I think there will be more significant changes required as there were a number of breaking changes, and I'm not really sure whether many ppl are using 1.x? I haven't scoped out the changes but I'd have to work through this document here 
  - https://github.com/aws/aws-sdk-js/blob/master/UPGRADING.md 

Let me know if we feel like this is a requirement here and I can start to scope this out.

Lastly it's worth pointing out that 3.x is being worked on and is coming at some point so that is probably something to keep an eye on as it looks to be a [complete rewrite in typescript](https://github.com/aws/aws-sdk-js-v3). We may want to restrict to >2, <3 in the meantime?

ericmustin replied:
updated to support >2.1, which ensures support before `.promise()` was introduced. didn't go all the way down to >2 because there were some bugs in 2.0.3 and earlier which make testing and general usage of the library very difficult, such as https://github.com/aws/aws-sdk-js/commit/a6c03caab714f4ddc2f6f731692d2c2ccc81444c

ericmustin commented:
@rochdev ok I think I've implemented or tried to give followup answers/comments on all the feedback, fwiw the tests are in a better place to review, not totally sure why they're flaking in the ci right now (they were passing earlier). I'm just starting to move things over to use the helper functions introduced recently (ty @bengl  these are great!) , have to add some testing support for `>=2` as discussed in other comment, and would like to add a few more tests for more specific cases, but general structure is good for a review at this point.

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,109 @@
+        analyticsSampler.sample(span, config.analytics)
+
+        // when passed a callback makeRequest will call `.send` on the AWS.Request
+        // before we have a chance to add our listeners, so we have to wrap
```
Is it possible to wrap the Request prototype so that we get to run first?

ericmustin replied:
I don't quite follow, do you mean, literally shim the entire `AWS.Request.prototype` or do you mean bind the `request` object before calling it? 

```
const boundRequest = tracer.scope().bind(span, request)
return boundRequest.call(
  this,
  operation,
  params,
  awsHelpers.wrapCallback(tracer, span, cb, childOf, config, serviceName)
)
```

Or am i shimming, `AWS.request.prototype.send`? Not totally clear what you mean/what the goal is here?

rochdev replied:
I mean shim the entire prototype. Then we wouldn't need 2 code paths since we'd be guaranteed to run first.

rochdev replied:
Just to be clear, by entire prototype I mean the method(s) where we need to run first, not every method.

ericmustin replied:
i went a level lower to shim `AWS.request.prototype.send` and `AWS.request.prototype.promise`. This also uncovered to me that min version that can be instrumented here is 2.1.35 (instead of 2.1.0) as there is a breaking change `<2.1.35` in the way the emit'ing of events occurs. This is from late 2015 so i think it's still a very reasonable default min version.

https://github.com/aws/aws-sdk-js/pull/629

Otherwise moving down a level lower simplifies the code paths quite a bit as the instrumentation can now always attach listeners before `.send()` is invoked and the state machine is kicked off.

ericmustin replied:
I encountered a slight hiccup here which is that when merely starting a span on `validate`, and finishing a span on `complete` events , the active scope was not being correctly picked up by nested integrations in `aws.request`. Specifically, that meant the `http`/`tcp`/`dns` integrations were associating with the parent span of `aws.request` (which would be an express span in the example below, and not `aws.request`. Here's a snapshot of this behaviour

![incorrect_child_parent_relationship](https://cl.ly/0ceab6d2d58a/%255B5e01f0b3703104804e12adfbb28aba2f%255D_incorrect_child_parent_relationship.png)

I've been able to correct this is by explicitly starting the span in the shim and then wrapping the invocation of the original `AWS.request.send()` in the `aws.request` span's activated scope. Since this is now a level lower and wrapping the `AWS.request.send` method explicitly, which is what kicks off the state machine being run until it's complete (instead of the first iteration which was wrapping  `aws.Service.makeRequest` , which merely instantiated the state machine but did not yet kicking off it's run)  this should still ensure the start time of the span is correct. Here's a snapshot of the updated behaviour:

![correct_child_parent_relationship](https://cl.ly/baf1bd66cf33/%255B6d0d3a49dfb26746e6540cef7729ca5c%255D_correct_child_parent_relationship.png)

I've also added tests for this behavior as well as tests to ensue that any callbacks passed in are associated with the parent scope and not the scope of `aws.request`

rochdev replied:
The new behavior is not correct as it seems every AWS request is a child of the previous request. To fix this, the scope when `send` was called should be restored in the callback.

ericmustin replied:
@rochdev my bad i included an older screenshot, i've since correct this behavior

updated: 

![nested properly](https://cl.ly/3bb4a6a2f92a/Image%2525202020-02-24%252520at%2525208.52.48%252520PM.png)

Here's the test app i used to generate this (it assumes localstack test docker container is running which should be listening on the correct host/port as long as the test env is set up appropriately)

https://gist.github.com/ericmustin/80636a955bb68f345059665580b936d2

Here's the LOC and also the test that demonstrates the callback is bound to the correct parent scope

LOC: https://github.com/DataDog/dd-trace-js/blob/486361f85e79d3f9870c5dfd2a9f3ae1cbb0fa6b/packages/datadog-plugin-aws-sdk/src/index.js#L34
Test: https://github.com/DataDog/dd-trace-js/blob/486361f85e79d3f9870c5dfd2a9f3ae1cbb0fa6b/packages/datadog-plugin-aws-sdk/test/index.spec.js#L1069

Again, my bad on the older screenshot, was rushing.

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,109 @@
+        // instrumenting those ensures requests w/out callback gets started/ended correctly
+        // https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L142
+        // https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L328-L332
+        boundAwsReq.on('validate', () => {
```
Can this fail? How is the failure reported?

ericmustin replied:
see: https://github.com/DataDog/dd-trace-js/pull/856#discussion_r382562374

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,109 @@
+          tracer.scope().activate(span)
+        })
+
+        boundAwsReq.on('complete', response => {
```
Can this fail? How is the failure reported?

ericmustin replied:
afaik, this is what the state machine looks like

```
  this.addState('validate', 'build', 'error', transition);
  this.addState('build', 'afterBuild', 'restart', transition);
  this.addState('afterBuild', 'sign', 'restart', transition);
  this.addState('sign', 'send', 'retry', transition);
  this.addState('retry', 'afterRetry', 'afterRetry', transition);
  this.addState('afterRetry', 'sign', 'error', transition);
  this.addState('send', 'validateResponse', 'retry', transition);
  this.addState('validateResponse', 'extractData', 'extractError', transition);
  this.addState('extractError', 'extractData', 'retry', transition);
  this.addState('extractData', 'success', 'retry', transition);
  this.addState('restart', 'build', 'error', transition);
  this.addState('success', 'complete', 'complete', transition);
  this.addState('error', 'complete', 'complete', transition);
  this.addState('complete', null, null, transition);
```

  - from: https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L44-L57

with `addState` being a function that accept the state, the success state, and the fail state:

  - `function addState(name, acceptState, failState, fn)`
  - from: https://github.com/aws/aws-sdk-js/blob/cc29728c1c4178969ebabe3bbe6b6f3159436394/lib/state_machine.js#L30

At each `transition` the new state is emitted

  - `self.emit(self._asm.currentState, ...`

  - from: https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L22

Which is what triggers our instrumentation's `validate` and `complete` callbacks

In the case `validate` fails, it would transition to `error`, which then transitions to `complete`. 

```
this.addState('validate', 'build', 'error', transition);
...
this.addState('error', 'complete', 'complete', transition);
```

`complete` receives a `response` object which includes `data` and `error` keys, this `response.error` key is what we pass into `.finish` and in `.finish` we check for the existence of any errors.

Basically all errors or successes within the state machine eventually get funneled to `complete` no matter what, and `complete` is the end of the request cycle.

As I understand it and from my tests, this means we should be correctly accounting for any errors, and not interfering with how they're handled.

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,109 @@
+        boundAwsReq.on('complete', response => {
+          if (!span) return
+
+          // response is same AWS.Response object
```
This comment seems redundant.

ericmustin replied:
agreed, removed

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,109 @@
+
+        // The state machine has garaunteed first and last events, validate / complete
+        // instrumenting those ensures requests w/out callback gets started/ended correctly
+        // https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L142
```
This code may change over time since we instrument so many versions.

ericmustin replied:
ok, i dont mind removing this for now, i looked into whether we could programmatically _always_ get the guaranteed first event in the case that they change this down the line, but looks like  it's just hardcoded ( https://github.com/aws/aws-sdk-js/blob/5608fd3b8f9f3b57b3c77e3a7cd701dfd0d4b070/lib/request.js#L328) 

removed comments, will just have to watch for any breaking changes here i think

rochdev replied:
This would be a huge breaking change, so I don't expect it to change unless a new major version is released.

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,109 @@
+        const awsReq = request.apply(this, arguments)
+        const boundAwsReq = tracer.scope().bind(awsReq, childOf)
+
+        // The state machine has garaunteed first and last events, validate / complete
```
This is implicit since we should always start/end at the guaranteed locations.

ericmustin replied:
makes sense, removed

ericmustin commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,206 @@
+      serviceName = serviceName.trim().replace(/\s/g, '')
+
+      if (serviceName.startsWith(prefix)) {
+        return `${serviceName.slice(0, prefix.length)}.${serviceName.slice(prefix.length)}`
```
understood, updated `my-service-name-aws-<serviceidentifier>`

ericmustin commented:
@rochdev ok tried to update everything i could / clarify on the latest round of feedback, i did have one question on your comment regarding wrapping Aws Request. I've also updated the tests to be a bit more robust (testing for a service we don't include in our case statements) and to handle <2.3 style code (without the .promise() function), along with updating the formatting/conventionns of the metadata as we discussed offline.

rochdev commented on index.d.ts:
```diff
@@ -633,6 +634,22 @@ declare namespace plugins {
+      /**
+       * Hook to execute just before the aws span finishes.
+       */
+      request?: (span?: opentracing.Span, response?: anyObject) => any;
```
Is the type definition for the Response available anywhere? If se we could copy it to provide the type.

ericmustin replied:
We do, i'm not particularly with how to do this in ts though, I just bulk copy/pasta this definition? there are some other aws specific types it relies on as well https://github.com/aws/aws-sdk-js/blob/5b51b4e3297f0f1f13d66eaff0c2e8ad5473de49/lib/response.d.ts#L3

rochdev replied:
Yes, unfortunately this is the only way to do it for libraries that don't have a separate package for the types.

rochdev replied:
As discussed, if this ends up being too complicated because of the way aws-sdk type definitions are exposed, let's leave `anyObject` for now.

rochdev commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    this.addRequestIdTag(span, response)
+    config.hooks.request(span, response)
```
Given the call to `this.addRequestIdTag(span, err)` in the `finish` helper, it's unclear if `addResponseTags` is called for errors. If that's not the case, then the hook will never be called.

ericmustin replied:
There's a weird edge case where errors will, in certain library versions, sometimes append the `requestId` to the `response.error` object itself instead of the `response` object, I will try to dig up the reference to this in their docs, but this is basically a defensive check for that. the service specific response tagging will be called for errors as well, as the `complete` event will always return the response, and the response will have a `response.error` key defined in the case of an error [ts link](
https://github.com/aws/aws-sdk-js/blob/5b51b4e3297f0f1f13d66eaff0c2e8ad5473de49/lib/request.d.ts#L129).

Either way since the tagging has been refactored to just use the base/service class structure (described here: https://github.com/DataDog/dd-trace-js/pull/856#discussion_r384422783) I've just removed this `addRequestIdTag`method and hardcoded the edge case check into the `finish` method

```
  finish (span, err) {
    if (err) {
      span.setTag('error', err)

      if (err.requestId) {
        span.addTags({ 'aws.response.request_id': err.requestId })
      }
    }

    span.finish()
  }
```

rochdev commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,172 @@
+    if (tags['resource.name']) return
+
+    Object.assign(tags, {
+      'resource.name': operation || 'Amazon'
```
I'd leave it blank by default, meaning remove the fallback, which will make it use `aws.request`.

ericmustin replied:
fair enough, removed

rochdev commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,172 @@
+
+  addDefaultTags (operation, tags) {
+    // defaults
+    if (tags['resource.name']) return
```
Why this check?

ericmustin replied:
Basically this was needed for the case of a service either:

1. we don't have a special convention for, or
2. a service we did have a convention for but was using malformed params or service method that wasn't specific enough to fit our conventions

since moving to the base/service class structure described here https://github.com/DataDog/dd-trace-js/pull/856#discussion_r384422783 this is now not necessary and has been removed

rochdev commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,172 @@
+    })
+  },
+
+  addDynamoDbTags (params, operation, tags) {
```
I think one way to organize this better, especially as we continue to improve this plugin, would be to split each service in its own file with a common interface, and then this can be made a lot more dynamic.

For example:

```js
const services = {
  dynamo: require('./services/dynamo') 
}

addServicesTags (span, response, serviceName) {
  const service = services[serviceName]
  if (operation && params && service) {
    service.addTags(params, operation, tags)
  }
}
```

This could even be improved to only lazy-load helpers for services that are needed. It will also make it a lot more readable when we add things like context propagation and so on.

ericmustin replied:
yea, agreed, makes sense. I've refactored the approach here to use classes in separate files with some base class behavior. I think it's much cleaner and just generally more extensible. I've held off on the lazy loading bit for now, per our discussion, because I'd rather kick the can on worrying about any perf impact of doing synchronous file loading (though it's almost certainly trivial)

Essentially instead of doing all these seperate methods for adding different tags and that big case statemment, 

 - theres a base class which generates default tags and assigns any additional tags generated by `_addServiceTags`, which is a no-op that returns an empty object on the base class, so basically

```
    const tags = Object.assign({<default tags>}, this._addServiceTags())
```


 - service classes which define _addServiceTags with their own service specific tagging logic
- a `getService` helper function which looks up and creates an instance of the service specific class based on the service name, and, in the case it doesnt exist, uses the base class 
- Then the only actual code in the helper is

```
  const service = getService(serviceName)
  service.addTags(span, params, operation, response)
```

Definitely think its much simpler and extensible going forward. there's one weird edge case where error's will have the requestId attached, but i've just hardcoded that additional check in the error handling code. anyway lmk what you think

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,90 @@
+        'aws.operation': this.operation,
+        'aws.region': this.service.config && this.service.config.region,
+        'aws.service': this.service.api && this.service.api.className,
+        'aws.url': this.service.endpoint && this.service.endpoint.href,
```
Will this be the same as the `http` plugin `http.url` value? If so, it can probably be omitted.

ericmustin replied:
yup it should be, removed

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,90 @@
+      const childOf = tracer.scope().active()
+      const baseTags = {
+        [Tags.SPAN_KIND]: 'client',
+        'span.type': 'http',
```
This should be `aws` if we support it in the backend, and otherwise it shouldn't be set.

ericmustin replied:
ok, removed. I don't think `aws` is supported, so removing this. Fwiw this was just copying the serverless convention, which does assign `dynamodb` it's own span.type, so i've left that bit in within the `DynamoDbService` class

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,90 @@
+
+      const span = tracer.startSpan('aws.request', {
+        childOf,
+        tags: baseTags
```
Small nit: this would be more readable if only called `tags`.

ericmustin replied:
updated

rochdev commented on packages/datadog-plugin-aws-sdk/src/index.js:
```diff
@@ -0,0 +1,90 @@
+    name: 'aws-sdk',
+    versions: ['>=2.3.0'],
+    patch (AWS, tracer, config) {
+      this.wrap(AWS.Request.prototype, 'promise', createWrapRequest(tracer, config))
```
Just to make sure, `promise` never calls `send`?

ericmustin replied:
No, not afaik from looking through the definition and through testing (the case we're worried about here is causing double patching?). The promise code is tremendously difficult to follow but this is basically the guts of it, basically our patching doesn't seem to pick up the `complete` event without wrapping `promise` specifically 
https://github.com/aws/aws-sdk-js/blob/38bf84c144281f696768e8c64500f2847fe6f298/lib/request.js#L779

rochdev commented on packages/datadog-plugin-aws-sdk/test/fixtures.js:
```diff
@@ -0,0 +1,110 @@
+'use strict'
```
Fixtures by service (different files) would be easier to work with.

ericmustin replied:
i've refactored this to split out the fixtures into their own files, lmk if the structure looks alright to you, should be easier to sort through now.

rochdev commented on packages/datadog-plugin-aws-sdk/test/index.spec.js:
```diff
@@ -0,0 +1,1137 @@
+'use strict'
```
Is there any way to not repeat the main tests for every service? I wonder if maybe a few of the it blocks could be extracted to a reusable function, at least for any generic tags/propagation/etc.

ericmustin replied:
Mocha is fighting me to the death on this, so trying to work on this still, the before/after setup/teardown does not want to play nice with abstracting some of these tests into a reusable function.

Should hopefully have something here by EOD

ericmustin replied:
So i've broken out there tests into as reusable of functions as I could, and ensured that i'm now properly testing each generic thing for each service, the drawback here is that these tests now take a very long time, they will probably time out on the ci, so i need to work on getting them sped up somehow

rochdev replied:
Nice, I think you actually pushed it even further than what I had in mind at that point 😄

rochdev commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,57 @@
+'use strict'
+
+const DEFAULT = 'default'
+const Services = {
```
Variables should be lowerCamelCase.

rochdev commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,57 @@
+
+function getService (serviceName) {
+  if (Services[serviceName]) {
+    return new Services[serviceName]()
```
If the variable was named with an uppercase first later because of ESLint, just setting another variable would fix it like `Service = services[name]`.

rochdev commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,57 @@
+
+    const params = response.request.params
+    const operation = response.request.operation
+    const service = getService(serviceName)
```
This creates a new instance on every request.

rochdev replied:
Maybe instead the `require` could be in `getServices` and then the services variable at the top would become:

```js
services = {
  s3: getService('s3')
}
```

ericmustin replied:
Ah, yea that's no good. I've updated this so there's now only 1 instance instantiated for each service. I also tried to introduce the pattern you suggested above, tho i believe this now introduces lazy-loading by wrapping the requires in a function (_which should be fine but just wanted to make sure I was thinking about it correctly_)? also, unless i assign the class that's returned by the `require` to an upper case constant , eslint yells at me,  
```
const services = {
  example_service_name: getService('example_service_name'),
}

function getService (serviceName) {
  const Service = require(`./services/${serviceName}`)
  return new Service()
}
```
, i thought this was proper formatting tho?

rochdev replied:
>  tho i believe this now introduces lazy-loading by wrapping the requires in a function

It doesn't because the function is called synchronously.

> also, unless i assign the class that's returned by the require to an upper case constant , eslint yells at me

That's fine since at that point the variable represents a class and not an object, so it should have an uppercase first character.

rochdev commented on packages/datadog-plugin-aws-sdk/src/helpers.js:
```diff
@@ -0,0 +1,57 @@
+  if (Services[serviceName]) {
+    return new Services[serviceName]()
+  } else {
+    return new Services[DEFAULT]()
```
Why not just do the base stuff outside of the class hierarchy? Then all classes simply have to implement the right interface.

ericmustin replied:
It feels like there are two types of service tags(and values): default and 'extra'. I wanted to make a clean way to ensure the 'default' tags are always present but the 'extra' tags takes precedence if present. Doing this via a "base" service object, and then adding/overwriting that objects' keys/values depending on the "extra" service specific stuff, _before_ calling span.addTags felt much cleaner then, say, calling span.addTags multiple times, or having to check the span's tags in and then conditionally call span.addTags, or having to generate and pass in a "default tags" object to the service method?

ericmustin replied:
after syncing, updated to just do the base calss work within `helpers.addServiceTags` and removed the class hierarchy

rochdev commented on packages/datadog-plugin-aws-sdk/src/services/cloudwatchlogs.js:
```diff
@@ -0,0 +1,19 @@
+
+const BaseService = require('./base')
+
+class CloudwatchLogsService extends BaseService {
```
`Service` in the name is redundant.

ericmustin replied:
k, i've updated these service related classes to just be the service name

rochdev commented on packages/datadog-plugin-aws-sdk/src/services/cloudwatchlogs.js:
```diff
@@ -0,0 +1,19 @@
+  _addServiceTags (params, operation, response) {
+    const tags = {}
+
+    // cloudwatach log group name
```
Comment is redundant since the variable is named exactly the same as the comment.

rochdev replied:
There are several other occurrences in other services as well.

ericmustin replied:
yup, removed basically all the comments in these services as either there's now an appropriate test for the behavior or the naming itself makes it redundant

rochdev commented on packages/datadog-plugin-aws-sdk/src/services/cloudwatchlogs.js:
```diff
@@ -0,0 +1,19 @@
+
+    return Object.assign(tags, {
+      'resource.name': `${operation} ${params.logGroupName}`,
+      'aws.cloudwatch_logs.log_group_name': params.logGroupName
```
`aws.cloudwatch.logs.log_group_name` would allow to group different CloudWatch product in the same object.

ericmustin replied:
makes sense, updated, tests updated as well

rochdev commented on packages/datadog-plugin-aws-sdk/test/index.spec.js:
```diff
@@ -0,0 +1,1084 @@
+  agent.wipe()
+}
+
+const baseSpecCheck = (done, agent, operation, serviceName, klass, fixture, key, metadata, config) => {
```
I'd say to move these to an external file since they make it difficult to see where the tests actually begin. Also, it's possible to move the entire `it` in the function as well to make the tests even cleaner.

rochdev replied:
It may even be possible to have a function that wraps all the others if every test uses the same ones.

ericmustin replied:
i've extracted these into a separate spec_helper file. However, basically anytime I extract anything at the size of an `it` block or larger into it's own re-usable function mocha executes things in the wrong order, so I think I would lean toward separating each service into it's own spec file if we feel the size here still makes this too large to reason about.

ericmustin replied:
Actually I think I was able to abstract this out a bit more, just had some scope issues that I had missed. Will push up an update here shortly

ericmustin replied:
Pushed up some nice little abstractions around the super repetitive chunks of tests for each service, should be  much easier to reason about now and extend to new services going forward

rochdev replied:
Did you solve the issue with `it` ordering? The way it works, it will always execute in the correct order unless `it` was not called in the right place, which shouldn't happen given the code is synchronous.

ericmustin replied:
Yes, the issue was that the `before` `after` hooks were behaving strangely when I removed the `it` blocks into their own helper function, but I was able to resolve the issue and things are now executing in the appropriate order.

marcghorayeb commented:
nice work 👍
